services/api/service.go

```diff
@@ -493,6 +527,84 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.CurrentCounter(),
+	})
+	if simErr != nil {
+		ignoreError := simErr.Error() == ErrBlockAlreadyKnown || simErr.Error() == ErrBlockRequiresReorg || strings.Contains(simErr.Error(), ErrMissingTrieNode)
+		if !ignoreError {
+			// Mark builder as non-optimistic.
+			opts.builder.status.IsOptimistic = false
+			log.WithError(simErr).Error("block validation failed")
+			return nil, simErr
+		}
+	}
+	if reqErr != nil {
+		log.WithError(reqErr).Error("block validation failed: request error")
+		return reqErr, nil
+	}
+	log.Info("block validation successful")
+	return nil, nil
+}
+
+func (api *RelayAPI) demoteBuilder(pubkey string, req *common.BuilderSubmitBlockRequest, simError error) {
+	builderEntry, ok := api.blockBuildersCache[pubkey]
+	if !ok {
+		api.log.Warnf("builder %v not in the builder cache", pubkey)
+		builderEntry = &blockBuilderCacheEntry{} //nolint:exhaustruct
+	}
+	newStatus := common.BuilderStatus{
+		IsHighPrio:    builderEntry.status.IsHighPrio,
+		IsBlacklisted: builderEntry.status.IsBlacklisted,
+		IsOptimistic:  false,
+	}
+	api.log.Infof("demoted builder, new status: %v", newStatus)
+	if err := api.db.SetBlockBuilderStatus(pubkey, newStatus, true); err != nil {
+		api.log.Error(fmt.Errorf("error setting builder: %v status: %w", pubkey, err))
+	}
+	// Write to demotions table.
+	api.log.WithFields(logrus.Fields{"builder_pubkey": pubkey}).Info("demoting builder")
+	if err := api.db.InsertBuilderDemotion(req, simError); err != nil {
+		api.log.WithError(err).WithFields(logrus.Fields{
+			"errorWritingDemotionToDB": true,
+			"bidTrace":                 req.Message,
+			"simError":                 simError,
+		}).Error("failed to save demotion to database")
+	}
+}
+
+// processOptimisticBlock is called on a new goroutine when a optimistic block
+// needs to be simulated.
+func (api *RelayAPI) processOptimisticBlock(opts blockSimOptions) {
+	api.optimisticBlocksInFlight.Add(1)
+	defer func() { api.optimisticBlocksInFlight.Sub(1) }()
+	api.optimisticBlocksWG.Add(1)
+	defer api.optimisticBlocksWG.Done()
+
+	ctx := context.Background()
+	builderPubkey := opts.req.BuilderPubkey().String()
+	opts.log.WithFields(logrus.Fields{
+		"builderPubkey": builderPubkey,
+		// NOTE: this value is just an estimate because many goroutines could be
+		// updating api.optimisticBlocksInFlight concurrently. Since we just use
+		// it for logging, it is not atomic to avoid the performance impact.
+		"optBlocksInFlight": api.optimisticBlocksInFlight,
+	}).Infof("simulating optimistic block with hash: %v", opts.req.BuilderSubmitBlockRequest.BlockHash())
+	reqErr, simErr := api.simulateBlock(ctx, opts)
+
+	if reqErr != nil || simErr != nil {
+		api.log.WithError(simErr).Error("block simulation failed in processOptimisticBlock, demoting builder")
```

Comment:
should this log with error level? error level is usually reserved for system errors, although on this one i'm on the fence

Reply:
changed to warn!